source/endpoints/ruleset.c

```diff
@@ -0,0 +1,878 @@
+            aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_UNSUPPORTED_RULESET);
+            goto error_cleanup;
+        }
+        ruleset->service_id = aws_string_new_from_cursor(allocator, &service_id_cur);
+    }
+
+    ruleset->parameters = aws_mem_acquire(allocator, sizeof(struct aws_hash_table));
+    AWS_ZERO_STRUCT(*ruleset->parameters);
+    aws_hash_table_init(
+        ruleset->parameters,
+        allocator,
+        20,
+        aws_hash_c_string,
+        aws_hash_callback_c_str_eq,
+        aws_hash_callback_string_destroy,
+        aws_hash_callback_endpoints_parameter_destroy);
+
+    struct aws_json_value *parameters_node =
+        aws_json_value_get_from_object(root, aws_byte_cursor_from_c_str("parameters"));
+    if (s_init_members_from_json(allocator, parameters_node, ruleset->parameters, s_on_parameter_key)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract parameters.");
+        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+        goto error_cleanup;
+    }
+
+    struct aws_json_value *rules_node = aws_json_value_get_from_object(root, aws_byte_cursor_from_c_str("rules"));
+    if (!aws_json_value_is_array(rules_node)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Unexpected type for rules node.");
+        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+        goto error_cleanup;
+    }
+    size_t num_rules = aws_json_get_array_size(rules_node);
+    aws_array_list_init_dynamic(&ruleset->rules, allocator, num_rules, sizeof(struct aws_endpoints_rule *));
+    if (s_init_array_from_json(allocator, rules_node, &ruleset->rules, s_on_rule_element)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract rules.");
+        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+        goto error_cleanup;
+    }
+
+    aws_json_value_destroy(root);
+    return AWS_OP_SUCCESS;
+
+error_cleanup:
+    aws_json_value_destroy(root);
+    return AWS_OP_ERR;
+}
+
+static void s_endpoints_ruleset_destroy(void *data) {
+    struct aws_endpoints_ruleset *ruleset = data;
+
+    aws_string_destroy(ruleset->version);
+    aws_string_destroy(ruleset->service_id);
+
+    if (ruleset->parameters) {
+        aws_hash_table_clean_up(ruleset->parameters);
+        aws_mem_release(ruleset->allocator, ruleset->parameters);
+    }
+
+    aws_array_list_deep_cleanup(&ruleset->rules, s_on_rule_array_element_destroy);
+
+    aws_mem_release(ruleset->allocator, ruleset);
+}
+
+struct aws_endpoints_ruleset *aws_endpoints_ruleset_new_from_string(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *ruleset_cur) {
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(ruleset_cur));
+
+    struct aws_endpoints_ruleset *ruleset = aws_mem_acquire(allocator, sizeof(struct aws_endpoints_ruleset));
+    AWS_ZERO_STRUCT(*ruleset);
```

Comment:
trivial: use `aws_mem_calloc()` to allocate zeroed-out memory
instead of `aws_mem_acquire(); AWS_ZERO_STRUCT()`

(search through the code for "AWS_ZERO_STRUCT" to find a lot more places you can do this)

Reply:
Whats the general guidance with acquire vs calloc? All of sdkutils is pretty much using acquire

Reply:
the general guidance is: use _calloc() if you're allocating a struct, because that avoids having uninitialized memory if you forget to initialize every single variable

use _acquire() if you're allocating something like the backing memory for an aws_byte_buf, where we're not even going to access the memory until something else has overwritten it